internal/charmstore/migrations.go

```diff
@@ -222,3 +227,96 @@ func populateWriteACL(db StoreDatabase) error {
 	logger.Infof("%d base entities updated", counter)
 	return nil
 }
+
+func populatePromulgatedEntities(db StoreDatabase) error {
+	baseEntities := db.BaseEntities()
+	entities := db.Entities()
+
+	// 1. Update All Base Entities to have a promulgated value of the correct form.
+	_, err := baseEntities.UpdateAll(
+		bson.D{{"promulgated", true}},
+		bson.D{{"$set", bson.D{{"promulgated", 1}}}},
+	)
+	if err != nil {
+		return errgo.Notef(err, "cannot set promulgated to initial value")
+	}
+	_, err = baseEntities.UpdateAll(
+		bson.D{{"$or", []interface{}{
+			bson.D{{"promulgated", false}},
+			bson.D{{"promulgated", bson.D{{"$exists", false}}}},
+		}}},
+		bson.D{{"$set", bson.D{{"promulgated", -1}}}},
+	)
+	if err != nil {
+		return errgo.Notef(err, "cannot set promulgated to initial value")
+	}
+
+	// 2. Update entities with users with their promulgated URL.
+	owners := make(map[string]string)
+	iter := entities.Find(bson.D{{"user", ""}}).Select(bson.D{
+		{"_id", 1},
+		{"baseurl", 1},
+		{"blobhash", 1},
+		{"extrainfo.bzr-owner", 1},
+	}).Sort("name", "revision").Iter()
+	var entity mongodoc.Entity
+	for iter.Next(&entity) {
+		var user string
+		if err := json.Unmarshal([]byte(entity.ExtraInfo["bzr-owner"]), &user); err != nil {
+			return errgo.Notef(err, "cannot unmarshal user from extra-info")
+		}
+		if user == "" {
+			return errgo.Newf("no user for %q", entity.URL)
+		}
+		owners[entity.URL.Name] = user
+		err := entities.Update(
+			bson.D{
```

Comment:
Do we have to worry about several entities with different revisions but the same blobhash here? If that happened, ISTM that this might cause a revision to be lost, as we'll be updating twice with the same query parameters, and the second update will change the same document. Perhaps this won't happen in practice, but maybe we could mitigate the possibility by only updating entries that have an unset promulgated-url field?